### ISSUE TITLE
[Autoscaling] RFC: EPP-predictor-based autoscaling

### ISSUE THREAD

**Issue by kaushikmitr** [llm-d/llm-d-autoscaling#1018]
# RFC: EPP-predictor-based autoscaling

## Summary

Adding a new `epp-saturation` analyzer that consumes the pool-level saturation
signal emitted by the gateway-api-inference-extension (EPP) latency detector,
instead of computing saturation from per-pod vLLM metrics.

**Code** (on my fork, not yet upstreamed):
- Analyzer: [`internal/engines/analyzers/epp_saturation/`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/tree/main/internal/engines/analyzers/epp_saturation)
- Engine wiring: [`internal/engines/saturation/engine_epp_saturation.go`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/blob/main/internal/engines/saturation/engine_epp_saturation.go)
- Query registration: [`internal/collector/registration/epp_saturation.go`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/blob/main/internal/collector/registration/epp_saturation.go)
- Deploy: [`deploy/deploy-epp-saturation.sh`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/blob/main/deploy/deploy-epp-saturation.sh) + [`deploy/README-epp-saturation.md`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/blob/main/deploy/README-epp-saturation.md)

This issue presents validation results from a GKE experiment and proposes a
few refinements based on what we observed.

## Key discussion points

Two questions I'd like community input on up front. They're the framing I
most want feedback on before going deeper into the proposal below.

### 1. How do we show this (or any analyzer) is actually working?

There's no agreed benchmark in this repo for evaluating an analyzer's quality.
The data I collected (Pearson correlation between signal and actual TTFT,
SLO violation rate, replica-count stability) is a reasonable first cut, but
it's ad-hoc, I made it up. Before investing in more analyzers, it would
be valuable to agree on a standard evaluation harness.

Some specific questions:

- **What defines "correct" behavior?** Is it (a) signal tracks actual
  latency, (b) replica recommendations keep SLO violations under X%,
  (c) total GPU cost minimized subject to SLO, (d) some composite? Each
  implies a different quality metric.
- **What workload patterns should we evaluate against?** Single-level
  steady state, ramps, bursts, bimodal prompt sizes, hot-key imbalance?
- **What's the "ground truth" baseline to compare against?** Manually
  tuned static replica count? Perfect oracle scaler? HPA on raw QPS?
- **Should evaluation be bundled into `test/e2e/`?** Today the suite tests
  correctness (does WVA produce a decision); it doesn't test quality (is
  the decision good). An e2e harness that runs a fixed workload and scores
  the analyzer output against the above metrics would let us compare
  analyzers apples-to-apples.

This proposal includes validation data for one run on one workload on one
cluster. It would be much more useful if there were a reusable harness that
anyone could point at any analyzer config.

### 2. Is the queueing-model analyzer already sufficient? What gaps does this address?

The existing queueing-model analyzer is already sophisticated: it learns
per-variant `(α, β, γ)` parameters online via a Kalman filter, uses
M/M/1/K queueing math to translate arrival rate into predicted TTFT/ITL,
and binary-searches for the max sustainable `λ` under explicit SLO
targets. That's strictly more principled than a single-signal approach.

So: **is adding the `epp-saturation` analyzer actually useful, or should
effort go into improving queueing-model instead?**

My case for adding it (open to being told I'm wrong):

- **Avoids duplicating work the EPP already does**. The EPP's latency
  detector already has a predictor trained on real traces; WVA's
  queueing-model tuner is learning the same hardware from fewer samples
  via Kalman. If the EPP signal is available, consuming it avoids
  maintaining two parallel models of the same thing.
- **Cold-start behavior**. Queueing-model needs real traffic for the
  Kalman filter to converge; until then it can't produce useful signals.
  The EPP signal is immediately usable (with a fallback probe input).
  This matters for bursty or low-baseline-traffic workloads.
- **Simpler mental model for operators**. One threshold (`scaleUpThreshold`)
  vs. learning-rate tuning, NIS outlier rejection, tuningEnabled toggles,
  TTFT/ITL SLO target splits, etc.

On the other hand, queueing-model has some things the current
`epp-saturation` implementation doesn't. All of which look **fixable
rather than fundamental** when examined closely:

- **Queueing-model has explicit, separate TTFT and ITL SLO targets.** The
  current `epp-saturation` analyzer reads a single pool-level score which
  is already `max(predictedTTFT/TTFTSLO, predictedTPOT/TPOTSLO)` for
  streaming mode. Useful, but collapses the two targets into one number.
  This is **not a fundamental gap**. The EPP already emits
  `inference_extension_latency_detector_endpoint_saturation` per-endpoint,
  and adding separate TTFT/TPOT pool signals is a small EPP change. WVA
  could then prefer per-target signals when available.
- **Queueing-model has redundant input signals** (per-pod queue + KV +
  TTFT + ITL observations). `epp-saturation` currently reads a single
  aggregate metric from the EPP. This is also fixable. The EPP already
  has per-endpoint saturation, and we could consume that for heterogeneous
  pools or cross-check against the aggregate. Single-signal is a current
  choice, not an inherent limitation.
- **Queueing-model can answer counterfactuals** ("what latency at λ' or
  N'?") via its physical M/M/1/K model. A saturation snapshot can't -
  but WVA isn't limited to consuming only the pre-computed saturation
  metric. It could **query the EPP's latency predictor directly** with
  any synthetic payload (any input tokens, any prefix cache score, any
  assumed per-endpoint state). Combined with a simple state-evolution
  model (how queue depth and KV cache usage scale with replica count
  under Poisson arrivals. The same math queueing-model uses), WVA
  could do its own counterfactual search against the EPP's already-trained
  predictor. That would replace the Kalman-filter-learned (α, β, γ) with
  the predictor's richer XGBoost features, while keeping the "search for
  max sustainable λ" logic.

So in principle, almost everything queueing-model does analytically can
also be built on top of the EPP predictor. And the EPP predictor has
access to more features than `(α, β, γ)` captures. The pieces exist
separately today; this issue is as much about deciding the direction as
about landing the specific code change.

I don't think the answer is "one or the other". But I'm genuinely unsure
where the line should be. Interested in maintainer/community perspective
on whether this analyzer earns its place alongside queueing-model, whether
the queueing-model's `(α, β, γ)` tuner should be replaced by calls to the
EPP predictor, or whether the effort is better spent elsewhere entirely.

## Motivation

The existing analyzers in WVA (V1/V2 saturation, queueing-model) all scrape
per-pod vLLM metrics and reason about capacity in WVA. This means:

- WVA needs to track each vLLM version's metric naming conventions
- WVA duplicates capacity logic that the EPP already implements
- The analyzers are **reactive**. They see the state that produced latency,
  not a prediction of what latency would be

The EPP's latency detector plugin (see
[saturation-detector-v2 branch](https://github.com/llm-d/gateway-api-inference-extension/tree/saturation-detector-v2))
already does the hard work: probe each endpoint with a synthetic representative
request, call a latency predictor sidecar, and compute
`saturation = predictedLatency / SLO` averaged across the pool. This is
**predictive**. It answers "would a request arriving now meet SLO?" rather
than "is the pool currently saturated?"

The `epp-saturation` analyzer consumes this pre-computed signal directly via
the Prometheus metric `inference_extension_latency_detector_pool_saturation`.

### The scaling formula

Given `S` = EPP saturation score and `N` = current replicas, WVA uses a
normalized capacity model:

```
supply       = 1.0                      (pool's full SLO budget)
demand       = S                        (saturation is demand/capacity)
perReplicaCap = 1/N                     (each replica contributes 1/N of supply)
```

Scaling decisions (with defaults `T_up = 0.85`, `T_down = 0.50`):

```
       ⎧ min(ceil(S × N / T_up),   maxReplicas)    if S > T_up    (scale up)
desired = ⎨ N                                       if T_down ≤ S ≤ T_up
       ⎩ max(ceil(S × N / T_down), minReplicas)    if S < T_down   (scale down)
```

The hysteresis band between `T_down` and `T_up` absorbs signal noise in the
normal operating range.

### Setup

- GKE cluster with 10× vLLM pods serving Qwen3-32B on H100-80GB
- EPP with `latency-detector` plugin + latency predictor sidecar
- SLOs (streaming mode): TTFT ≤ 3s, TPOT ≤ 100ms
- WVA in observe-only mode (HPA disabled), `maxReplicas=20`
- 55-minute experiment with load ramping 0 → 7.7 QPS and back
- Data captured every 60s from Prometheus

### Results

<img width="1544" height="1537" alt="Image" src="https://github.com/user-attachments/assets/9c756c48-e5f3-4745-9b94-6a0208e80d24" />

#### Signal quality

| Metric | Value |
|---|---|
| Pearson(saturation, actual_TTFT) | **+0.89** |
| Pearson(saturation, actual_TPOT) | +0.53 (TPOT was not binding) |
| Time range | 12:19 – 13:14 (55 min) |
| TPOT SLO violations | **0/54 cycles** |
| TTFT SLO violations | 20/54 cycles (37%). During the overload phase |

Under this workload the signal is strongly correlated with the binding SLO
(TTFT). The predictor did not clamp at 1.0. It reached **saturation=6.15**
when actual TTFT was ~137s (46× SLO), which is the right order of magnitude.

#### Scaling behavior

| Phase | Time | QPS | Saturation | WVA rec |
|---|---|---|---|---|
| Startup | 12:19–12:24 | 0 | 3–6 (stale) | 10 |
| Light load | 12:25–12:55 | 0.4–5.5 | 0.43–0.82 | 9–10 |
| Overload | 12:55–13:14 | 4.9–7.7 | 1.0–6.15 | 11–20 |

- **Predictive**: at 12:55 (sat=0.82, TTFT≈1.9s), WVA was still at 10. At
  12:56 (sat=1.04, TTFT≈8s), WVA jumped to 12. The signal led actual latency
  by ~1 minute.
- **Proportional**: as saturation climbed 1→2→3→5→6, WVA recommended
  12→18→20(capped). Wanted 45 at peak, correctly capped by `maxReplicas=20`.
- **No direction flipping** during the ramp. Smooth single-direction progression.

### Caveat: the linearity assumption may be wrong

The formula assumes `new_saturation ≈ old_saturation × N / N'`. I.e. that
saturation is **inversely linear in replica count**. This is the load
argument: if you double the number of pods, per-pod load halves, so
saturation halves.

**This is almost certainly not accurate in general.** Real queueing
latency is super-linear near saturation (`W ∝ 1/(1-ρ)`), so:

- **Near saturation**: a small increase in replica count should cause
  a *disproportionately large* drop in saturation (better than linear).
  Which means the formula **under-scales** when the system is overloaded
 . It recommends fewer replicas than would be needed for linearity, but
  fewer replicas *actually suffice* because queueing dynamics help.
- **Far from saturation**: at low load, adding or removing replicas
  barely changes per-pod latency (the pods are mostly idle). So the
  formula will keep reducing replicas on scale-down until it hits the
  point where removing more actually matters. Potentially overshooting.
- **The EPP signal itself isn't purely a function of load**: it also
  depends on the predictor's input profile (from the input-profile-tracker),
  which can shift independently of replica count. The formula treats
  saturation as a pure function of `λ/N`, but it isn't.

Worked example from the overload phase:

- Peak overload: sat=3.77, actual TTFT≈44s, N=10
- Formula's recommendation: `ceil(3.77 × 10 / 0.85) = 45` replicas
  (capped at `maxReplicas=20`)
- If linearity held, at N=20 we'd expect saturation ≈ 3.77 × 10/20 ≈ 1.88
- What actually happened at 13:08 (N=11, sat=0.91, TTFT≈1.7s): the
  system recovered faster than linearity predicted, 11 pods brought
  TTFT to 1.7s (well under SLO) while the formula would have predicted
  saturation ≈ 3.77 × 10/11 ≈ 3.43.

So the formula **over-recommended scale-up** at peak overload in a way
that worked out fine because reality was non-linear in our favor. But
the same non-linearity could go the other way in different workloads or
at different operating points, so **I don't want to claim the formula is
"validated" by one experiment**. It's more honest to say: the formula
is a heuristic that happens to work in the data we have, and its
theoretical justification (linearity) is only approximately true.

This is one of the strongest arguments for the
[counterfactual approach discussed above](#2-is-the-queueing-model-analyzer-already-sufficient-what-gaps-does-this-address):
if WVA can query the EPP's latency predictor directly with hypothetical
state, we can stop relying on a linearity assumption and use the richer
prediction instead.

### 1. Signal smoothing (EMA)

The EPP saturation signal updates every probe cycle based on recent traffic
samples, and its inputs (per-endpoint queue depth, KV cache, probe input
profile percentile) can shift cycle-to-cycle. Using the raw signal directly
in the scaling math makes replica recommendations sensitive to single-cycle
swings.

**Approach**: apply an EMA in the analyzer before the value feeds the
scaling math. Commit `c7d53ad`. Default `alpha=0.3` gives an effective
window ~17s at 5s reconcile cycle. Tunable via `smoothingAlpha` in
`wva-saturation-scaling-config`, or set to 1.0 to disable.

Live logs show the damping in action (`raw=0.567 → smoothed=0.534`), and
boundary oscillation at the scale-down threshold is bounded to ±1 replica
rather than catastrophic drops.

**Rationale for smoothing in WVA vs elsewhere**: smoothing is a scaling-policy
decision. Different consumers of the EPP metric (dashboards, admission,
routing) want the raw instantaneous signal. WVA is the right layer because
it's where the signal is translated into replica counts, and operators can
tune the smoothing to match their workload's cost/responsiveness preference.

### 2. `maxReplicas` cap behavior

During the overload phase, the formula wanted ~45 replicas but was clamped
at `maxReplicas=20`. This is the intended guardrail. Operators should
set `maxReplicas` to match their GPU budget and trust the cap. The issue
is that operators currently can't easily distinguish "system is fine at 20"
from "WVA wanted more but was blocked."

**Proposal**: Emit a `wva_scale_capped{variant=...}` gauge (1 when the
recommendation was clamped by `maxReplicas`, 0 otherwise), and add a
`ScalingCapped` condition on the VA status with the uncapped recommendation
in the message. This is a small, additive change that gives operators the
visibility they need to tune `maxReplicas`.

### 3. Smoothing observability

Right now the smoothed saturation is only visible in logs. For operators
to tune `smoothingAlpha` effectively against their workload, it's more
useful as a Prometheus metric alongside the raw value.

**Proposal**: Emit `wva_epp_saturation_raw{variant=...}` and
`wva_epp_saturation_smoothed{variant=...}` as gauges.

### 4. Fallback behavior when the EPP signal is unavailable

Unlike analyzers that combine queue depth and KV cache, this one relies on
a single signal from the EPP's latency detector. If the detector fails (for
instance, the predictor sidecar is down or a probe gets stuck), the analyzer
has no alternative capacity signal.

WVA already has a safety-net path for stale/missing metrics that preserves
the last known desired replica count and marks `MetricsAvailable: False`.
This analyzer should use the same path. Flagging here to confirm the
behavior is consistent and documented.

## Related commits

All on my fork ([`kaushikmitr/llm-d-workload-variant-autoscaler`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler)):

- [`5eae9a3`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/commit/5eae9a3), analyzer implementation
- [`7a34ec8`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/commit/7a34ec8), GMP PodMonitoring for WVA metrics
- [`c7d53ad`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/commit/c7d53ad), EMA smoothing
- [`a86d0f8`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/commit/a86d0f8), vLLM ITL metric name fix (needed for queueing-model, noticed during comparison)
- [`7676e35`](https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/commit/7676e35), honor `OptimizationInterval` config (needed to tune reconcile speed)

## Data

Raw CSV data from the GKE experiment (happy to share on request or attach to this issue):
- `saturation.csv`, `wva_desired_replicas.csv`, `median_ttft.csv`, `median_tpot.csv`, `qps.csv`

**Comment by kaushikmitr** [comment]
[Doc](https://docs.google.com/document/d/17CmMkw2wTol-Nkiy-EWWmzAUSm-K8lQCc3xlr_U0Ck0/edit?tab=t.0) for further discussions

**Comment by lionelvillard** [comment]
@kaushikmitr I moved this issue to the v0.8.0 milestone. Let us know if you think the timeline is too aggressive and/or if you want to break down this RFC into smaller tasks.

**Comment by kaushikmitr** [comment]
@asm582 @lionelvillard @ev-shindin @atantawi The PR:  #1376  is ready (i will break it down into smaller PRs for easier review, most of the files are related to docs). Here is the key result, wanted to share this first to make sure looks good to you. 

https://github.com/kaushikmitr/llm-d-workload-variant-autoscaler/blob/epp-saturation-observability/docs/developer-guide/epp-saturation-benchmark.md

**Comment by kaushikmitr** [comment]
as discussed with @asm582 this can be implemented directly in KEDA. Created this PR:
http://github.com/llm-d/llm-d/pull/2011
